Sidebar: Add a shared component for the inserter and list view

scruffian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Creates a new `TabbedPanel` component that is used by both the inserter and the list view. 

## Why?
This will ensure visual and behaviour consistency between these two panels. In time we should also update the block inspector and global styles to use the component too.

## How?
- Create a new `TabbedPanel` component
- Refactor shared code to use the new component

## Testing Instructions
0. Open the Site Editor
1. Open the block inserter and check that the tabs behave the same way as in trunk
2. Open the list view and check that the tabs behave the same way as in trunk

### Testing Instructions for Keyboard
Also check that the keyboard interactions are the same as in trunk.

Note: There is ongoing discussion about the best way to handle keyboard interactions with these panels. This PR aims to consolidate these panels to use the same code and approach. When we update the approach to keyboard interactions then we'll only need to do it in one place for all panels.

## Screenshots or screencast <!-- if applicable -->

|  Before | After  |
|---|---|
| <img width="682" alt="Screenshot 2024-06-17 at 12 39 27" src="https://github.com/WordPress/gutenberg/assets/275961/46bb6a7b-52e6-4379-8e8f-4362e2e77b05">| <img width="362" alt="Screenshot 2024-07-02 at 13 14 18" src="https://github.com/WordPress/gutenberg/assets/275961/08f64618-06c8-473a-9fe5-c80db3223942"> |

| <img width="376" alt="Screenshot 2024-06-17 at 12 41 30" src="https://github.com/WordPress/gutenberg/assets/275961/cc910429-209c-402f-8e7c-1fcb46d6ff82"> |  <img width="373" alt="Screenshot 2024-07-02 at 13 14 30" src="https://github.com/WordPress/gutenberg/assets/275961/c8340930-65d1-49ba-a4eb-f3a97daa250d"> |

The visual differences are:
 - The blue border under the tabs are now the same, which means the one in the inserter is now slightly wider
 - The tabs in the list view are very slightly wider

scruffian committed:
remove the old panels

scruffian committed:
move styles

scruffian committed:
rename classes

scruffian committed:
trying to fix the overflow

scruffian committed:
remove will-change: transform

scruffian committed:
update docs

scruffian committed:
add the correct markup for the list view

scruffian committed:
remove padding

scruffian committed:
remove unsued code

scruffian committed:
remove unused code

scruffian committed:
remove unused comment

scruffian committed:
move styles across

scruffian committed:
readd height

scruffian committed:
remove more code

DaniGuardiola commented:
Inner styles should not be overriden for `@wordpress/components`, in general. If we want a version of tabs that stretches to fill its container, that should be requested and implemented (if agreed upon) in the component itself, as a variant. Style overrides are very fragile and make maintenance of the underlying component pretty painful.

Separately, there are some valid arguments against these full-width styles, like https://github.com/WordPress/gutenberg/pull/61974#issuecomment-2149719264

While I think a component like this that unifies styles is a good addition (without having looked too deep into the details), I think the conversation in #61974 should be resolved first before committing to one or the other. If full-width is the decision, then there should be work to implement those styles in `Tabs` before merging this. cc @mirka @ciampo

scruffian commented:
We are already overriding these styles in the individual components, so this doesn't break that convention any more than it's already been broken, it just centralizes that code.

DaniGuardiola commented:
@scruffian I think that's reasonable. Please note that the first example in the description has an intentional design detail where the tabs indicator overlaps the lower border. If we unify all of these (and the ones you mention for the future), we should also make a unified decision regarding that detail. cc @WordPress/gutenberg-design

jasmussen commented:
Good catch, @DaniGuardiola, that is definitely intentinoal, and is about ensuring legibility of the active state, so that's an important detail to retain.

jeryj commented on packages/block-editor/src/components/tabbed-sidebar/index.js:
```diff
@@ -0,0 +1,71 @@
+					<Button
+						className="block-editor-tabbed-sidebar__close-button"
+						icon={ closeSmall }
+						label={ __( 'Close block inserter' ) }
```
This will need to be dynamic, as the list view close button tooltip also says "Close block inserter"

![](https://github.com/WordPress/gutenberg/assets/967608/f2e1b7ea-19ca-4d73-8719-bb23ddcb9480)

jeryj commented on packages/block-editor/src/components/index.js:
```diff
@@ -162,6 +162,7 @@ export { default as __experimentalPublishDateTimePicker } from './publish-date-t
 export { default as __experimentalInspectorPopoverHeader } from './inspector-popover-header';
 export { default as BlockPopover } from './block-popover';
 export { useBlockEditingMode } from './block-editing-mode';
+export { default as TabbedSidebar } from './tabbed-sidebar';
```
This is a helpful refactor to work towards unified sidebars, so I don't want to block this progress. That said, I think it will be important to not combine the content of the sidebar with the interaction of the sidebar wrapper. I see the eventual goal as having a shared sidebar component (or another name that does not reference the visual position of the component) with a close button. The content within it can be whatever it needs to be. 

This will be possible if we move to a header title for each sidebar, as proposed in https://github.com/WordPress/gutenberg/pull/61553

jeryj commented on packages/block-editor/README.md:
```diff
@@ -804,6 +804,10 @@ _Related_
 
 -   <https://github.com/WordPress/gutenberg/blob/HEAD/packages/data/README.md#registerStore>
 
+### TabbedSidebar
```
Should this be a public component?

jeryj commented on packages/block-editor/src/components/tabbed-sidebar/README.md:
```diff
@@ -0,0 +1,76 @@
+import { TabbedSidebar } from '@wordpress/components';
+
+const MyTabbedSidebar = () => (
+	<TabbedSidebar
+		tabs={ [
+			{
+				name: 'slug-1',
+				title: _x( 'Title 1', 'context' ),
+				panel: <PanelContents>,
+				panelRef: useRef('an-optional-ref'),
+			},
+			{
+				name: 'slug-2',
+				title: _x( 'Title 2', 'context' ),
+				panel: <PanelContents />,
+			},
+		] }
+		onClose={ onClickCloseButton }
+		onSelect={ onSelectTab }
+		defaultTabId="slug-1"
+		ref={ tabsRef }
+	/>
```
To elaborate on the above, I think this should eventually look more like:

```
<StickyPanel title="Block Inserter" onClose={ onClickCloseButton }>
  <Tabs>
    ...
   </Tabs>
</StickyPanel>
```
Name tbd 😬 😅

DaniGuardiola replied:
Agree on this. We should try to keep things composable to avoid ["YAP" syndrome](https://kyleshevlin.com/quit-your-yapping/).

jeryj reviewed (commented):
Love all the red of removed code :) 

With the zoom-out pattern experiment enabled, the patterns list is not positioned correctly. 
![Patterns sidebar shown to the right of a 300px wide gap of white space](https://github.com/WordPress/gutenberg/assets/967608/748a82dc-c4a8-40ce-a928-d6f2b8a63ab0)

DaniGuardiola commented:
Just wanna say two things:

- Great work, thank you for doing this!
- This is a great opportunity to make sure this component is rock-solid in terms of a11y. Let's please ensure that we're very confident about this aspect before merging. This includes any nuance about ARIA roles and interactivity (like @afercia mentions in this comment: https://github.com/WordPress/gutenberg/pull/62343#issuecomment-2178084836), and thorough testing, as well as a sufficient strategy for integration with the rest of the app in a way that doesn't break a11y. While this work doesn't necessarily need to involve me, I'm happy to engage if it helps.

jeryj commented:
@DaniGuardiola - to improve the accessibility, we'd need design agreement for [adding a header with title and close button to all the sidebars](https://github.com/WordPress/gutenberg/pull/61553). I would love if this could get momentum. It has agreement and support from folks on the WordPress Accessibility team.

DaniGuardiola commented:
@jeryj love that! This PR is also a great chance to add that as a follow-up, and get it going in all sidebars.

youknowriad commented:
Based on the discussion on the linked PR, it seems a consensus is going to be hard to reach for that design. Can we instead focus on the unification part in this PR while we continue iterating on the design separately.

DaniGuardiola commented:
I agree with @youknowriad, let's leave the visual design discussion separate from this PR.

Let's still make sure this component is accessible to all users, please :)

jeryj commented:
Agreed :) I think any changes to the styles should be a follow-up. 

After this is merged, can this tabbed sidebar component be applied to the Settings panel as well?

scruffian committed:
fix the position of patterns

scruffian committed:
use different labels for each close button

scruffian commented:
> @scruffian I think that's reasonable. Please note that the first example in the description has an intentional design detail where the tabs indicator overlaps the border below it, as can be seen in the "before" image. If we unify all of these (and the ones you mention for the future), we should also make a unified decision regarding that detail. cc

Thanks. I have moved the indicator so it overlaps the border and updated the screenshots.

scruffian committed:
Make TabbedSidebar private

scruffian commented:
I have updated this based on the feedback above. I have made it a locked component to give us time to iterate on it.

> After this is merged, can this tabbed sidebar component be applied to the Settings panel as well?

Definitely

jeryj reviewed (approved):
I think this is better than what we have since it consolidates the code more and can also be applied to the Settings sidebar. However, I do have concerns over how the Tabs are tied to a close button that controls the sidebar outside of it. It feels like two unrelated things are being forced together. But, pragmatically, this is better than what we have and the component is private so we should be able to refactor it when the time comes.

scruffian commented:
I'm going to merge this so that we can continue to align the various panels to use this component and ensure that improvments are made to all at once. I am very happy for others to jump in and make improvments - for example making the component more composable etc.
